rust/src/asn1/mod.rs

```diff
@@ -0,0 +1,447 @@
+/// ad_ptr must be a valid object obtained using `rs_detect_asn1_parse`
+///
+/// Returns 1 if any of the options match, 0 if not
+#[no_mangle]
+pub unsafe extern "C" fn rs_asn1_checks(
+    ptr: *const Asn1,
+    ad_ptr: *const DetectAsn1Data,
+) -> u8 {
+    if ptr.is_null() || ad_ptr.is_null() {
+        return 0;
+    }
+
+    let asn1 = &*ptr;
+    let ad = &*ad_ptr;
+
+    match asn1.check(ad) {
+        Ok(Some(_check)) => 1,
+        Ok(None) => 0,
+        Err(e) => {
+            SCLogError!("error during asn1 checks: {}", e.to_string());
+            0
+        }
+    }
+}
+
+impl From<std::num::TryFromIntError> for Asn1DecodeError {
+    fn from(_e: std::num::TryFromIntError) -> Asn1DecodeError {
+        Asn1DecodeError::InvalidKeywordParameter
+    }
+}
+
+impl From<nom::Err<der_parser::error::BerError>> for Asn1DecodeError {
+    fn from(e: nom::Err<der_parser::error::BerError>) -> Asn1DecodeError {
+        match e {
+            nom::Err::Incomplete(_) => Asn1DecodeError::InvalidLength,
+            nom::Err::Error(e) | nom::Err::Failure(e) => match e {
+                BerError::BerTypeError => Asn1DecodeError::BerTypeError,
+                BerError::BerValueError => Asn1DecodeError::BerValueError,
+                BerError::InvalidTag => Asn1DecodeError::InvalidTag,
+                BerError::InvalidClass => Asn1DecodeError::InvalidClass,
+                BerError::InvalidLength => Asn1DecodeError::InvalidLength,
+                BerError::ConstructExpected => Asn1DecodeError::ConstructExpected,
+                BerError::ConstructUnexpected => Asn1DecodeError::ConstructUnexpected,
+                BerError::IntegerTooLarge => Asn1DecodeError::IntegerTooLarge,
+                BerError::BerMaxDepth => Asn1DecodeError::BerMaxDepth,
+                BerError::ObjectTooShort => Asn1DecodeError::ObjectTooShort,
+                BerError::DerConstraintFailed => Asn1DecodeError::DerConstraintFailed,
+                BerError::UnknownTag => Asn1DecodeError::UnknownTag,
+                BerError::Unsupported => Asn1DecodeError::Unsupported,
+                _ => Asn1DecodeError::InvalidStructure,
+            },
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use test_case::test_case;
+
+    // Example from the specification X.690-0207 Appendix A.3
+    static ASN1_A3: &[u8] = b"\x60\x81\x85\x61\x10\x1A\x04John\x1A\x01 \
+                    P\x1A\x05Smith\xA0\x0A\x1A\x08Director \
+                    \x42\x01\x33\xA1\x0A\x43\x0819710917 \
+                    \xA2\x12\x61\x10\x1A\x04Mary\x1A\x01T\x1A\x05 \
+                    Smith\xA3\x42\x31\x1F\x61\x11\x1A\x05Ralph\x1A\x01 \
+                    T\x1A\x05Smith\xA0\x0A\x43\x0819571111 \
+                    \x31\x1F\x61\x11\x1A\x05Susan\x1A\x01B\x1A\x05 \
+                    Jones\xA0\x0A\x43\x0819590717";
+
+    /// Ensure that the checks work when they should
+    #[test_case("oversize_length 132 absolute_offset 0", ASN1_A3, DetectAsn1Data {
```

Comment:
(not a comment) didn't know this crate, looks interesting!

Reply:
Very useful for "parameterized" tests, repetitive tests where the input changes but the test code does not